source/opt/local_ssa_elim_pass.h

```diff
@@ -0,0 +1,254 @@
+  // This assures correct depth first search in the presence of early 
+  // returns and kills. If the successor vector contain duplicates
+  // if the merge block, they are safely ignored by DFS.
+  void ComputeStructuredSuccessors(ir::Function* func);
+
+  // Compute structured block order for |func| into |structuredOrder|. This
+  // order has the property that dominators come before all blocks they
+  // dominate and merge blocks come after all blocks that are in the control
+  // constructs of their header.
+  void ComputeStructuredOrder(ir::Function* func,
+      std::list<ir::BasicBlock*>* order);
+
+  // Return true if loop header block
+  bool IsLoopHeader(ir::BasicBlock* block_ptr) const;
+
+  // Initialize SSA map for block with single predecessor
+  void SSABlockInitSinglePred(ir::BasicBlock* block_ptr);
+
+  // Return true if variable is loaded after the label
+  bool IsLiveAfter(uint32_t var_id, uint32_t label) const;
+
+  // Initialize SSA map for loop header block by merging SSA maps 
+  // from all predecessors. If any value ids differ for any variable
+  // across predecessors, create a phi function and use that value id
+  // for the variable in the new SSA map. Assumes all predecessors have
+  // been visited by SSARewrite except the back edge. Use a dummy value
+  // in the phi for the back edge until the back edge block is visited
+  // and patch the phi value then.
+  void SSABlockInitLoopHeader(std::list<ir::BasicBlock*>::iterator block_itr);
+
+  // Initialize SSA map for selection merge block by merging SSA Maps
+  // from all predecessors. If any value ids differ for any variable
+  // across predecessors, create a phi function and use that value id
+  // for the variable in the new SSA map. Assumes all predecessors have
+  // been visited by SSARewrite.
+  void SSABlockInitSelectMerge(ir::BasicBlock* block_ptr);
+
+  // Initialize the label2SSA map entry for a block. Insert phi instructions
+  // into block when necessary. All predecessor blocks must have been
+  // visited by SSARewrite except for backedges.
+  void SSABlockInit(std::list<ir::BasicBlock*>::iterator block_itr);
+
+  // Return undef in function for type. Create and insert an undef after the
+  // first non-variable in the function if it doesn't already exist. Add
+  // undef to function undef map.
+  uint32_t Type2Undef(uint32_t type_id);
+
+  // Patch phis in loop header block now that the map is complete for the
+  // backedge predecessor. Specifically, for each phi, find the value
+  // corresponding to the backedge predecessor. That contains the variable id
+  // that this phi corresponds to. Change this phi operand to the the value
+  // which corresponds to that variable in the predecessor map.
+  void PatchPhis(uint32_t header_id, uint32_t back_id);
+
+  // Return true if all extensions in this module are allowed by this pass.
+  // Currently, no extensions are allowed. glsl_std_450 extended instructions
```

Comment:
Yeah. I just didn't want people to think we weren't handling glsl_std_450 instructions, because we are. I will work on that line.